### PR TITLE
Tree array access.

### PR DESCRIPTION
**Pull request by jeromekelleher**
WIP for #1299

Trickier than I thought because of the circular references between the array and the Tree object, which owns the memory.

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Looks like this is working well. I think it's probably a good idea to add the ``flags_array`` and ``time_array`` to the tree class also, since these are things we'll often be using in low-level algorithms, and it's a faff (and currently copy-ful, semantics-wise) to do this via the tree sequence tables arrays.

Should we go ahead and merge this much, and try it out for a bit before documenting, or go ahead and document it now and be done with it? I doubt we can change much here, it's pretty much a done-deal I think.

@benjeffery, any thoughts?

**Review by benjeffery — APPROVED** [review]
This is awesome stuff, couple of small comments.

I wonder about the `_array` naming though. Would `left_children`, `parents` etc. fit more with our existing names as we don't put type information in names normally?

**Comment by benjeffery** [comment]
> Should we go ahead and merge this much, and try it out for a bit before documenting, or go ahead and document it now and be done with it? I doubt we can change much here, it's pretty much a done-deal I think.

> @benjeffery, any thoughts?

Lets merge this, then file issues under 0.3.6 for `time_array` and `flags_array` and the docs. I'm happy to pick those up if I get to them before you.

**Comment by jeromekelleher** [comment]
> I wonder about the _array naming though. Would left_children, parents etc. fit more with our existing names as we don't put type information in names normally?

It's a bit awkward, I agree, but I'm not sure there's much choice. If we were starting again I'd get rid of the ``parent(u)`` etc functions and use these arrays instead. But, that ship has sailed, so ``parent_array`` seems like the least bad option. I don't want to use ``parents`` because this will be break with the convention that array names are singular. Also, I think it would be confusing if we had ``tree.left_children[u]`` and ``tree.left_child(u)`` giving the same thing.

**Comment by jeromekelleher** [comment]
(Thanks for the review btw, will update later)

File: python/tests/test_highlevel.py [hunk 1]
```diff
@@ -2692,6 +2692,39 @@ def test_interval(self):
...
+            assert tree.right_sib(u) == tree.right_sib_array[u]
+
+    def test_tree_arrays(self):
+        ts = msprime.simulate(10, recombination_rate=1, random_seed=1)
+        assert ts.num_trees > 1
+        for tree in ts.trees():
+            self.verify_tree_arrays(tree)
+
+    @pytest.mark.parametrize(
+        "array", ["parent", "left_child", "right_child", "left_sib", "right_sib"]
```
[the remark this answers is not in the document]
Comment (jeromekelleher) [reply python/tests/test_highlevel.py:2716]: good call

**Comment by jeromekelleher** [comment]
OK, updated and documented. @petrelharp, fancy taking a quick look?

(I finally figured out how to make repeated bits of text in RST: this would come in handy for the stats API!)

File: docs/python-api.rst [hunk 2]
```diff
@@ -1,3 +1,11 @@
+.. |tree_array_warning| replace:: This is a high-performance interface which
```
Comment (petrelharp) [inline docs/python-api.rst:1]: woah nice

Comment (jeromekelleher) [reply docs/python-api.rst:1]: Yeah!

**Comment by petrelharp** [comment]
I did look at this, btw - looks good.

**Pull request merged**